### PR TITLE
New module: Add Openstack Tempest run module (cloud/openstack/os_tempest_run)

### PR DESCRIPTION
**Pull request by TalShafir**
##### SUMMARY
New module that lets the user run Tempest

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
os_tempest_run

##### ANSIBLE VERSION
```
ansible 2.4.0 (module/os_tempest_run c7c81835cb) last updated 2017/05/01 21:11:35 (GMT +300)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/tshafir/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/tshafir/ansible/lib/ansible
  executable location = /home/tshafir/ansible/bin/ansible
  python version = 2.7.5 (default, Aug 18 2016, 15:58:25) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

### CODE REVIEW

**Comment by TalShafir** [comment]
The bot didn't tag anyone again.
@emonty @shrews @juliakreger @j2sol @rcarrillocruz @thingee

ready_for_review

**Comment by TalShafir** [comment]
bot_status

File: lib/ansible/modules/cloud/openstack/os_tempest_run.py
Comment (gundalow) [inline lib/ansible/modules/cloud/openstack/os_tempest_run.py]: The `I(workspace)` will turn into italics, which I'm not sure if that's what you want in this case

https://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#formatting-options

File: lib/ansible/modules/cloud/openstack/os_tempest_run.py
Comment (gundalow) [inline lib/ansible/modules/cloud/openstack/os_tempest_run.py]: If you use the `mutually_exclusive` function you can remove this line

File: lib/ansible/modules/cloud/openstack/os_tempest_run.py
Comment (gundalow) [inline lib/ansible/modules/cloud/openstack/os_tempest_run.py]: As per Ansible best practice, please use `- name: Run all tests with default number of workers` then remove the `- ` from in front of the module name

http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#examples-block

File: lib/ansible/modules/cloud/openstack/os_tempest_run.py
Comment (gundalow) [inline lib/ansible/modules/cloud/openstack/os_tempest_run.py]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
mutually_exclusive
required_together
required_one_of
require_if
```


Should `whitelist_file` and `blacklist_file` be `mutually_exclusive`
`mutually_exclusive=(('blacklist_file', 'whitelist_file'),),`

File: lib/ansible/modules/cloud/openstack/os_tempest_run.py
Comment (gundalow) [inline lib/ansible/modules/cloud/openstack/os_tempest_run.py]: I wonder if it might be worth saying `Run OpenStack Tempest` to make it a little easier for people to find via a search engine.

File: lib/ansible/modules/cloud/openstack/os_tempest_run.py
Comment (gundalow) [inline lib/ansible/modules/cloud/openstack/os_tempest_run.py]: pyyhon2 -> python2

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/openstack/os_tempest_run.py
Comment (TalShafir) [reply lib/ansible/modules/cloud/openstack/os_tempest_run.py]: The documentation says "I() for option names" but I think I'll change that to C(workspace) to make it more visible.
edit: now that I think about it, it should be C(workspace) as it is the option value and not the name.

File: lib/ansible/modules/cloud/openstack/os_tempest_run.py
Comment (TalShafir) [reply lib/ansible/modules/cloud/openstack/os_tempest_run.py]: Thanks, I fixed it.
I originally wrote it after looking at the pip module as an example (https://docs.ansible.com/ansible/pip_module.html#examples), might be nice to add it as an issue for something like 'First Timers Only' or 'YourFirstPR'.

**Comment by TalShafir** [comment]
Thanks for the review @gundalow 
ready_for_review

**Comment by TalShafir** [comment]
ready_for_review
bot_status

File: lib/ansible/modules/cloud/openstack/os_tempest_run.py
Comment (gundalow) [reply lib/ansible/modules/cloud/openstack/os_tempest_run.py]: That's a good idea.

Where in the existing documentation/examples did you look?

I'm wondering where I can put/link to the example so people will find it.

**Comment by TalShafir** [comment]
ready_for_review

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
@TalShafir Thanks for the prompt updates.

I'll let someone that understands OpenStack do the final review, so over to @emonty @shrews @juliakreger @j2sol @rcarrillocruz @thingee

File: lib/ansible/modules/cloud/openstack/os_tempest_run.py
Comment (TalShafir) [reply lib/ansible/modules/cloud/openstack/os_tempest_run.py]: In the documentation website: https://docs.ansible.com/ansible/pip_module.html#examples
And in the module implementation starts at: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/language/pip.py#L148

**Comment by TalShafir** [comment]
@emonty @Shrews @juliakreger @j2sol @rcarrillocruz @Thingee
A kind reminder, waiting for your reviews.
Thanks!

**Comment by Shrews** [comment]
Hi!  Thanks for the new module. However, I don't feel like tempest modules are a good fit here. The current set of modules are geared toward *using* OpenStack services. Modules for *testing* the services seems like they should be kept separate. Most consumers of the core Ansible OpenStack cloud modules are only going to be interested in the former, not the latter.

And on a more personal level, as a core maintainer for these modules, I don't want to have to learn the tempest testing environment in order to support these going forward. I just have no interest in it.

I'd like to get more feedback from possibly @emonty, @j2sol, or @rcarrillocruz and get their thoughts on this.

**Comment by omgjlk** [comment]
I agree with @Shrews

**Comment by mnaser** [comment]
I agree with @omgjlk and @Shrews.  I would recommend the following role to run Tempest if you'd like

https://github.com/openstack/openstack-ansible-os_tempest

@emonty do you feel the same as well?

**Comment by sshnaidm** [comment]
Thanks for submitting patch for Openstack Ansible modules!
We moved Openstack Ansible modules to Openstack repositories.
Next patches should be submitted not with Ansible Github but with
Openstack Gerrit: https://review.opendev.org/#/q/project:openstack/ansible-collections-openstack
Please submit your code there from now.
Thanks for your contribution and sorry for inconvienience.

**Comment by Akasurde** [comment]
Closing as per above.

**Pull request closed without merging**
